src/main/java/org/apache/maven/plugins/shade/resource/SisuIndexResourceTransformer.java

```diff
@@ -0,0 +1,162 @@
+import java.util.List;
+import java.util.jar.JarEntry;
+import java.util.jar.JarOutputStream;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.maven.plugins.shade.relocation.Relocator;
+
+/**
+ * Resources transformer that relocates classes in {@code META-INF/sisu/javax.inject.Named} and appends resources
+ * into a single resource.
+ *
+ * @since 3.3.0
+ */
+public class SisuIndexResourceTransformer
+    extends AbstractCompatibilityTransformer
+{
+    private static final String SISU_INDEX_PATH = "META-INF/sisu/javax.inject.Named";
+
+    private ServiceStream serviceStream;
+
+    private List<Relocator> relocators;
+
+    private long time = Long.MIN_VALUE;
+
+    @Override
+    public boolean canTransformResource( final String resource )
+    {
+        return resource.equals( SISU_INDEX_PATH );
+    }
+
+    @Override
+    public void processResource( final String resource,
+                                 final InputStream is,
+                                 final List<Relocator> relocators,
+                                 long time ) throws IOException
+    {
+        if ( serviceStream == null )
+        {
+            serviceStream = new ServiceStream();
+        }
+
+        final String content = IOUtils.toString( is, StandardCharsets.UTF_8 );
+        StringReader reader = new StringReader( content );
+        BufferedReader lineReader = new BufferedReader( reader );
+        String line;
+        while ( ( line = lineReader.readLine() ) != null )
+        {
+            String relContent = line;
+            for ( Relocator relocator : relocators )
+            {
+                if ( relocator.canRelocateClass( relContent ) )
+                {
+                    relContent = relocator.applyToSourceContent( relContent );
+                }
+            }
+            serviceStream.append( relContent + "\n" );
+        }
+
+        if ( this.relocators == null )
+        {
+            this.relocators = relocators;
+        }
+
+        if ( time > this.time )
+        {
+            this.time = time;        
+        }
+    }
+
+    @Override
+    public boolean hasTransformedResource()
+    {
+        return serviceStream != null;
+    }
+
+    @Override
+    public void modifyOutputStream( final JarOutputStream jos )
+        throws IOException
+    {
+        ServiceStream data = serviceStream;
+
+        JarEntry jarEntry = new JarEntry( SISU_INDEX_PATH );
+        jarEntry.setTime( time );
+        jos.putNextEntry( jarEntry );
+
+
+        // read the content of service file for candidate classes for relocation.
+        // Specification requires that this file is encoded in UTF-8.
+        Writer writer = new OutputStreamWriter( jos, StandardCharsets.UTF_8 );
+        InputStreamReader streamReader = new InputStreamReader( data.toInputStream() );
+        BufferedReader reader = new BufferedReader( streamReader );
+        String className;
+
+        while ( ( className = reader.readLine() ) != null )
```

Comment:
commons IO probably has this logic for copying streams already

Reply:
fixed